File: internal/pkg/pinyin/token.go
package pinyin

import (
	"strings"
)

type Token struct {
	syllable string
	tones    Tones
}

func (t Token) String() string {
	var sb strings.Builder
	sb.WriteString(t.syllable)
	sb.WriteByte('-')
	for r := t.tones.Range(); r.Next(); {
		sb.WriteByte(byte('0' + r.Value()))
	}
	return sb.String()
}

func (t Token) Alts() Alts {
	result := Alts{}
	for r := t.tones.Range(); r.Next(); {
		result = append(result,
			Word{newPinyinFromSyllableAndTone(t.syllable, r.Value())})
	}
	return result
}

File: electron/src/pinyin/Token.ts
import Pinyin from './Pinyin';
import Tones from './Tones';
import Word from './Word';

export default class Token {
	constructor(
		public readonly syllable: string,
		public readonly tones?:   Tones,
	) {}

	get string(): string {
		return `${this.syllable}(${this.tones ? this.tones.string : ''})`
	}

	*[Symbol.iterator](): Iterator<Word> {
		const syllable = this.syllable;
		for (const tone of this.tones??[]) {
			yield new Word([new Pinyin({syllable, tone})]);
		}
	}
}
